transport/tlscommon/tls.go

```diff
@@ -0,0 +1,246 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package tlscommon
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+const logSelector = "tls"
+
+// LoadCertificate will load a certificate from disk and return a tls.Certificate or error
+func LoadCertificate(config *CertificateConfig) (*tls.Certificate, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	certificate := config.Certificate
+	key := config.Key
+	if certificate == "" {
+		return nil, nil
+	}
+
+	log := logp.NewLogger(logSelector)
+
+	certPEM, err := ReadPEMFile(log, certificate, config.Passphrase)
+	if err != nil {
+		log.Errorf("Failed reading certificate file %v: %+v", certificate, err)
+		return nil, fmt.Errorf("%w %v", err, certificate)
+	}
+
+	keyPEM, err := ReadPEMFile(log, key, config.Passphrase)
+	if err != nil {
+		log.Errorf("Failed reading key file %v: %+v", key, err)
+		return nil, fmt.Errorf("%w %v", err, key)
+	}
+
+	cert, err := tls.X509KeyPair(certPEM, keyPEM)
+	if err != nil {
+		log.Errorf("Failed loading client certificate %+v", err)
+		return nil, err
+	}
+
+	log.Debugf("Loading certificate: %v and key %v", certificate, key)
+	return &cert, nil
+}
+
+// ReadPEMFile reads a PEM formatted string either from disk or passed as a plain text starting with a "-"
+// and decrypt it with the provided password and  return the raw content.
+func ReadPEMFile(log *logp.Logger, s, passphrase string) ([]byte, error) {
+	pass := []byte(passphrase)
+	var blocks []*pem.Block
+
+	r, err := NewPEMReader(s)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	content, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	for len(content) > 0 {
+		var block *pem.Block
+
+		block, content = pem.Decode(content)
+		if block == nil {
+			if len(blocks) == 0 {
+				return nil, errors.New("no pem file")
+			}
+			break
+		}
+
+		if x509.IsEncryptedPEMBlock(block) { //nolint: staticcheck // deprecated, we have to get rid of it
```

Comment:
Should we have an issue for these deprecated x509 functions?

CC @belimawr whose looking at improving the TLS code already.

Reply:
It seems [PEM encryption as specified in RFC 1423is insecure by design](https://pkg.go.dev/crypto/x509#IsEncryptedPEMBlock) rather than this specific function. So it's a bigger discussion as it means removing an functionality.

Anyway it's interesting discussion because in our case the key and the encrypted data live together...